### PR TITLE
feat: adding Vault TLS client authentication

### PR DESCRIPTION
**Pull request by ricardoptcosta**
Part of issue #161 

Pull request for early review. Tests are still to be done.

### CODE REVIEW

File: apis/externalsecrets/v1alpha1/secretstore_vault_types.go [hunk 1]
```diff
@@ -161,3 +168,20 @@ type VaultJwtAuth struct {
...
+// method, with the role name and token stored in a Kubernetes Secret resource.
+type VaultCertAuth struct {
+	// ClientCert is a certificate to authenticate using the Cert Vault
+	// authentication method
+	// +optional
+	ClientCert string `json:"clientCert"`
+
+	// CACert is the Vault TLS Listener CA certificate
+	// +optional
+	CACert string `json:"caCert"`
```
Comment (moolen) [inline apis/externalsecrets/v1alpha1/secretstore_vault_types.go:182]: I think this is already covered by vault.CABundle here:

https://github.com/external-secrets/external-secrets/blob/7de9dc98c516dd104a28d670de0d10989711c46d/pkg/provider/vault/vault.go#L230-L238

File: pkg/provider/vault/vault.go [hunk 2]
```diff
@@ -530,3 +551,101 @@ func (v *client) requestTokenWithJwtAuth(ctx context.Context, client Client, jwt
...
+	caCertPath, err := getCertPath(certAuth.CACert, "ca.crt")
+	if err != nil {
+		return "", fmt.Errorf(errGetCertPath, err)
+	}
+
+	tlscfg := vault.TLSConfig{
+		ClientCert: clientCertPath,
+		ClientKey:  clientKeyPath,
+		CACert:     caCertPath,
+	}
```
Comment (moolen) [inline pkg/provider/vault/vault.go:581]: I think we can skip storing the certificates on disk and modify the `http.Client`'s TLSClientConfig directly (like the CABundle implementation above). That's what the `cfg.ConfigureTLS()` does under the hood for you.

https://github.com/hashicorp/vault/blob/api/v1.1.1/api/client.go#L241-L246

File: apis/externalsecrets/v1alpha1/secretstore_vault_types.go [hunk 3]
```diff
@@ -161,3 +168,20 @@ type VaultJwtAuth struct {
...
 	SecretRef esmeta.SecretKeySelector `json:"secretRef,omitempty"`
 }
+
+// VaultJwtAuth authenticates with Vault using the JWT/OIDC authentication
+// method, with the role name and token stored in a Kubernetes Secret resource.
+type VaultCertAuth struct {
+	// ClientCert is a certificate to authenticate using the Cert Vault
+	// authentication method
+	// +optional
+	ClientCert string `json:"clientCert"`
```
Comment (moolen) [inline apis/externalsecrets/v1alpha1/secretstore_vault_types.go:178]: The client cert is not sensitive information, however we could pull that out of a `Kind=Secret` aswell like the private key. Ideally a user has a `type=tls` secret. wdyt?

File: apis/externalsecrets/v1alpha1/secretstore_vault_types.go (hunk 1 above)
Comment (ricardoptcosta) [reply apis/externalsecrets/v1alpha1/secretstore_vault_types.go:182]: Thank you! I wasn't sure that would be kept in memory, but it is.

File: apis/externalsecrets/v1alpha1/secretstore_vault_types.go (hunk 3 above)
Comment (ricardoptcosta) [reply apis/externalsecrets/v1alpha1/secretstore_vault_types.go]: Good idea

File: pkg/provider/vault/vault.go (hunk 2 above)
Comment (ricardoptcosta) [reply pkg/provider/vault/vault.go]: 100%

**Comment by ricardoptcosta** [comment]
Created issue (#280) to implement unit tests.

**Comment by moolen** [comment]
@ricardoptcosta is this ready to be reviewed?

**Comment by moolen** [comment]
LGTM! I want to do manual testing to figure out how to set up vault with mTLS.

**Pull request merged**